### ISSUE TITLE
I dont know how to use the HTTPS proxy  with slimerjs and casperjs and PAC file

### ISSUE THREAD

**Issue by instagramlover** [laurentj/slimerjs#638]
### versions 

```
- SlimerJS: 
Innophi SlimerJS 0.10.3, Copyright 2012-2017 Laurent Jouanneau & Innophi
/usr/local/bin/slimerjs
```

```
- Firefox: 
Mozilla Firefox 52.2.0
/usr/bin/firefox
```

```
-casperjs
/usr/local/bin/casperjs
CasperJS version 1.1.4 at  using phantomjs version 2.1.1
```

```
- Operating system: 
centos 6.8
```

### Steps to reproduce the issue
I want to use Opera HTTPS proxy , which i use to use with firefox with PAC file loaded to firefox
The Pac File is 

`http://beshoo.com/pac.js`

```
function FindProxyForURL(url, host) {
	return 'HTTPS us.opera-proxy.net:443';
}
```

And the Auth information is : 

user : `XXXXXXX`
Password : `YYYYYYYY`

Now i have this js file 
```
var casper = require('casper').create({
    verbose: true,
    logLevel: "debug",
    pageSettings: {
        userAgent: 'Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/37.0.2062.120 Safari/537.36'
    }
});
        casper.start('https://www.instaranker.com/ip.php', function() {
                console.log(this.page.content)

        });

        casper.run(function() {
                this.echo('Finished with success!');
                casper.exit();
        });
```

I am calling this file via this command line : 

`casperjs --ssl-protocol=any ./test2.js   --debug=true --engine=slimerjs --proxy='http://beshoo.com/pac.js' --proxy-type='config-url' --proxy-auth='XXXXXXXX:YYYYYYYYY'`

Now i assume everything is correct since i can use this PAC file on my PC with firefox , a popup for username and password appear , once i provide username password it works.

Now with slimerjs, i think things gos different : i have this log 
 
### Actual results:
```
root@server [/home/mybeshoo/www/brows]# xvfb-run casperjs --ssl-protocol=any ./test2.js  --debug=true --engine=slimerjs --proxy='http://beshoo.com/pac.js' --proxy-type='config-url' --proxy-auth='XXXXXXXX:YYYYYYYYY'
Xlib:  extension "RANDR" missing on display ":99".
JavaScript strict warning: resource://gre/modules/TelemetryEnvironment.jsm, line 1131: ReferenceError: reference to undefined property "nsIShellService"
2017-08-05T02:38:29.901Z [DEBUG] Gecko version: 52.2.0
2017-08-05T02:38:29.901Z [DEBUG] script args: /home/mybeshoo/www/brows/casperjs/bin/bootstrap.js --casper-path=/home/mybeshoo/www/brows/casperjs --cli ./test2.js
2017-08-05T02:38:29.902Z [DEBUG] Configuration: proxy-type=config-url
2017-08-05T02:38:29.902Z [DEBUG] Configuration: proxy=http://beshoo.com/pac.js
2017-08-05T02:38:29.902Z [DEBUG] Configuration: proxy-auth=XXXXXXXX:YYYYYYYYY
2017-08-05T02:38:29.902Z [DEBUG] Configuration: debug=true
2017-08-05T02:38:29.902Z [DEBUG] Configuration: ssl-protocol=-2
2017-08-05T02:38:29.903Z [DEBUG] Configuration: Script=/home/mybeshoo/www/brows/casperjs/bin/bootstrap.js
2017-08-05T02:38:29.903Z [DEBUG] Configuration: workingDirectory=/home/mybeshoo/public_html/brows
JavaScript strict warning: coffee-scripts.js, line 8: SyntaxError: test for equality (==) mistyped as assignment (=)?
JavaScript strict warning: coffee-scripts.js, line 8: SyntaxError: test for equality (==) mistyped as assignment (=)?
JavaScript strict warning: coffee-scripts.js, line 8: SyntaxError: test for equality (==) mistyped as assignment (=)?
JavaScript strict warning: coffee-scripts.js, line 8: SyntaxError: test for equality (==) mistyped as assignment (=)?
JavaScript strict warning: coffee-scripts.js, line 8: SyntaxError: test for equality (==) mistyped as assignment (=)?
JavaScript strict warning: coffee-scripts.js, line 8: SyntaxError: test for equality (==) mistyped as assignment (=)?
JavaScript strict warning: coffee-scripts.js, line 8: SyntaxError: test for equality (==) mistyped as assignment (=)?
JavaScript strict warning: coffee-scripts.js, line 8: SyntaxError: test for equality (==) mistyped as assignment (=)?
JavaScript strict warning: coffee-scripts.js, line 8: SyntaxError: test for equality (==) mistyped as assignment (=)?
JavaScript strict warning: coffee-scripts.js, line 8: SyntaxError: test for equality (==) mistyped as assignment (=)?
JavaScript strict warning: coffee-scripts.js, line 8: SyntaxError: test for equality (==) mistyped as assignment (=)?
JavaScript strict warning: coffee-scripts.js, line 8: SyntaxError: test for equality (==) mistyped as assignment (=)?
JavaScript strict warning: coffee-scripts.js, line 8: SyntaxError: test for equality (==) mistyped as assignment (=)?
JavaScript strict warning: coffee-scripts.js, line 8: SyntaxError: test for equality (==) mistyped as assignment (=)?
JavaScript strict warning: coffee-scripts.js, line 8: SyntaxError: test for equality (==) mistyped as assignment (=)?
JavaScript strict warning: coffee-scripts.js, line 8: SyntaxError: test for equality (==) mistyped as assignment (=)?
JavaScript strict warning: coffee-scripts.js, line 8: SyntaxError: test for equality (==) mistyped as assignment (=)?
JavaScript strict warning: coffee-scripts.js, line 8: SyntaxError: test for equality (==) mistyped as assignment (=)?
JavaScript strict warning: coffee-scripts.js, line 8: SyntaxError: test for equality (==) mistyped as assignment (=)?
JavaScript strict warning: coffee-scripts.js, line 8: SyntaxError: test for equality (==) mistyped as assignment (=)?
JavaScript strict warning: coffee-scripts.js, line 8: ReferenceError: reference to undefined property "fs"
[info] [phantom] Starting...
[info] [phantom] Running suite: 2 steps
[debug] [phantom] opening url: https://www.instaranker.com/ip.php, HTTP GET
2017-08-05T02:38:35.082Z [DEBUG] webpage: openUrl https://www.instaranker.com/ip.php conf:{operation: "get", data: undefined, }
[debug] [phantom] Navigation requested: url=https://www.instaranker.com/ip.php, type=Undefined, willNavigate=true, isMainFrame=true
2017-08-05T02:38:35.095Z [DEBUG] network: main request starting - https://www.instaranker.com/ip.php flags:START,IS_REQ,IS_DOC,IS_NET,IS_WIN,
2017-08-05T02:38:35.163Z [DEBUG] network: resource request #1 started: GET - https://www.instaranker.com/ip.php flags=DOCUMENT_URI, INITIAL_DOCUMENT_URI,
JavaScript strict warning: resource://gre/modules/ProfileAge.jsm, line 202: ReferenceError: reference to undefined property "reset"
2017-08-05T02:38:37.413Z [DEBUG] network: status change for https://www.instaranker.com/ip.php (2152398858): Waiting for www.instaranker.com…
JavaScript error: resource://gre/components/nsLoginManagerPrompter.js, line 1411: TypeError: win.gBrowser is undefined
JavaScript error: , line 0: uncaught exception: 2147500033
2017-08-05T02:38:37.540Z [DEBUG] network: resource #1 response 'start': https://www.instaranker.com/ip.php flags=DOCUMENT_URI, INITIAL_DOCUMENT_URI,
2017-08-05T02:38:37.542Z [DEBUG] network: resource #1 response in error: #101 - The connection to the proxy server was refused
2017-08-05T02:38:37.542Z [DEBUG] network: resource #1 response end status: https://www.instaranker.com/ip.php
2017-08-05T02:38:37.543Z [DEBUG] network: resource #1 response in error (3): 101 - The connection to the proxy server was refused
2017-08-05T02:38:37.544Z [DEBUG] network: main request https://www.instaranker.com/ip.php flags:STOP,IS_REQ,
2017-08-05T02:38:37.544Z [DEBUG] network: main request: transfer done
2017-08-05T02:38:37.545Z [DEBUG] network: main request https://www.instaranker.com/ip.php flags:STOP,IS_DOC,
2017-08-05T02:38:37.545Z [DEBUG] network: main request: ignored state
2017-08-05T02:38:37.546Z [DEBUG] network: main request https://www.instaranker.com/ip.php flags:STOP,IS_NET,IS_WIN,
2017-08-05T02:38:37.546Z [DEBUG] network: main request: is loaded
JavaScript strict warning: chrome://global/content/bindings/browser.xml, line 389: ReferenceError: reference to undefined property "localName"
2017-08-05T02:38:37.562Z [DEBUG] network: request ignored. main page uri not started yet - jar:file:///usr/lib64/firefox/omni.ja!/chrome/toolkit/skin/classic/global/netError.css flags:START,IS_REQ,
[warning] [phantom] Loading resource failed with status=fail (HTTP 407): https://www.instaranker.com/ip.php
[debug] [phantom] Successfully injected Casper client-side utilities
[info] [phantom] Step anonymous 2/2 https://www.instaranker.com/ip.php (HTTP 407)
<!DOCTYPE html [
  <!ENTITY % htmlDTD
    PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
    "DTD/xhtml1-strict.dtd">
  %htmlDTD;
  <!ENTITY % netErrorAppDTD
    SYSTEM "chrome://global/locale/netErrorApp.dtd">
  %netErrorAppDTD;
  <!ENTITY % netErrorDTD
    SYSTEM "chrome://global/locale/netError.dtd">
  %netErrorDTD;
  <!ENTITY % globalDTD
    SYSTEM "chrome://global/locale/global.dtd">
  %globalDTD;
]>
<!-- This Source Code Form is subject to the terms of the Mozilla Public
   - License, v. 2.0. If a copy of the MPL was not distributed with this
   - file, You can obtain one at http://mozilla.org/MPL/2.0/. -->
<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <title>Page Load Error</title>
    <link rel="stylesheet" href="chrome://global/skin/netError.css" type="text/css" media="all">
    <!-- If the location of the favicon is changed here, the FAVICON_ERRORPAGE_URL symbol in
         toolkit/components/places/src/nsFaviconService.h should be updated. -->
    <link rel="icon" type="image/png" id="favicon" href="chrome://global/skin/icons/warning-16.png">

    <script type="application/javascript"><![CDATA[
      // Error url MUST be formatted like this:
      //   moz-neterror:page?e=error&u=url&d=desc
      //
      // or optionally, to specify an alternate CSS class to allow for
      // custom styling and favicon:
      //
      //   moz-neterror:page?e=error&u=url&s=classname&d=desc

      // Note that this file uses document.documentURI to get
      // the URL (with the format from above). This is because
      // document.location.href gets the current URI off the docshell,
      // which is the URL displayed in the location bar, i.e.
      // the URI that the user attempted to load.

      function getErrorCode()
      {
        var url = document.documentURI;
        var error = url.search(/e\=/);
        var duffUrl = url.search(/\&u\=/);
        return decodeURIComponent(url.slice(error + 2, duffUrl));
      }

      function getCSSClass()
      {
        var url = document.documentURI;
        var matches = url.match(/s\=([^&]+)\&/);
        // s is optional, if no match just return nothing
        if (!matches || matches.length < 2)
          return "";

        // parenthetical match is the second entry
        return decodeURIComponent(matches[1]);
      }

      function getDescription()
      {
        var url = document.documentURI;
        var desc = url.search(/d\=/);

        // desc == -1 if not found; if so, return an empty string
        // instead of what would turn out to be portions of the URI
        if (desc == -1)
          return "";

        return decodeURIComponent(url.slice(desc + 2));
      }

      function retryThis(buttonEl)
      {
        // Note: The application may wish to handle switching off "offline mode"
        // before this event handler runs, but using a capturing event handler.

        // Session history has the URL of the page that failed
        // to load, not the one of the error page. So, just call
        // reload(), which will also repost POST data correctly.
        try {
          location.reload();
        } catch (e) {
          // We probably tried to reload a URI that caused an exception to
          // occur;  e.g. a nonexistent file.
        }

        buttonEl.disabled = true;
      }

      function initPage()
      {
        var err = getErrorCode();

        // if it's an unknown error or there's no title or description
        // defined, get the generic message
        var errTitle = document.getElementById("et_" + err);
        var errDesc  = document.getElementById("ed_" + err);
        if (!errTitle || !errDesc)
        {
          errTitle = document.getElementById("et_generic");
          errDesc  = document.getElementById("ed_generic");
        }

        var title = document.getElementById("errorTitleText");
        if (title)
        {
          title.parentNode.replaceChild(errTitle, title);
          // change id to the replaced child's id so styling works
          errTitle.id = "errorTitleText";
        }

        var sd = document.getElementById("errorShortDescText");
        if (sd)
          sd.textContent = getDescription();

        var ld = document.getElementById("errorLongDesc");
        if (ld)
        {
          ld.parentNode.replaceChild(errDesc, ld);
          // change id to the replaced child's id so styling works
          errDesc.id = "errorLongDesc";
        }

        // remove undisplayed errors to avoid bug 39098
        var errContainer = document.getElementById("errorContainer");
        errContainer.parentNode.removeChild(errContainer);

        var className = getCSSClass();
        if (className && className != "expertBadCert") {
          // Associate a CSS class with the root of the page, if one was passed in,
          // to allow custom styling.
          // Not "expertBadCert" though, don't want to deal with the favicon
          document.documentElement.className = className;

          // Also, if they specified a CSS class, they must supply their own
          // favicon.  In order to trigger the browser to repaint though, we
          // need to remove/add the link element.
          var favicon = document.getElementById("favicon");
          var faviconParent = favicon.parentNode;
          faviconParent.removeChild(favicon);
          favicon.setAttribute("href", "chrome://global/skin/icons/" + className + "_favicon.png");
          faviconParent.appendChild(favicon);
        }
        if (className == "expertBadCert") {
          showSecuritySection();
        }

        if (err == "remoteXUL") {
          // Remove the "Try again" button for remote XUL errors given that
          // it is useless.
          document.getElementById("errorTryAgain").style.display = "none";
        }

        if (err == "cspBlocked") {
          // Remove the "Try again" button for CSP violations, since it's
          // almost certainly useless. (Bug 553180)
          document.getElementById("errorTryAgain").style.display = "none";
        }

        if (err == "nssBadCert") {
          // Remove the "Try again" button for security exceptions, since it's
          // almost certainly useless.
          document.getElementById("errorTryAgain").style.display = "none";
          document.getElementById("errorPageContainer").setAttribute("class", "certerror");
          addDomainErrorLink();
        }
        else {
          // Remove the override block for non-certificate errors.  CSS-hiding
          // isn't good enough here, because of bug 39098
          var secOverride = document.getElementById("securityOverrideDiv");
          secOverride.parentNode.removeChild(secOverride);
        }

        if (err == "inadequateSecurityError") {
          // Remove the "Try again" button for HTTP/2 inadequate security as it
          // is useless.
          document.getElementById("errorTryAgain").style.display = "none";

          var container = document.getElementById("errorLongDesc");
          for (var span of container.querySelectorAll("span.hostname")) {
            span.textContent = document.location.hostname;
          }
        }
      }

      function showSecuritySection() {
        // Swap link out, content in
        document.getElementById('securityOverrideContent').style.display = '';
        document.getElementById('securityOverrideLink').style.display = 'none';
      }

      /* In the case of SSL error pages about domain mismatch, see if
         we can hyperlink the user to the correct site.  We don't want
         to do this generically since it allows MitM attacks to redirect
         users to a site under attacker control, but in certain cases
         it is safe (and helpful!) to do so.  Bug 402210
      */
      function addDomainErrorLink() {
        // Rather than textContent, we need to treat description as HTML
        var sd = document.getElementById("errorShortDescText");
        if (sd) {
          var desc = getDescription();

          // sanitize description text - see bug 441169

          // First, find the index of the <a> tag we care about, being careful not to
          // use an over-greedy regex
          var re = /<a id="cert_domain_link" title="([^"]+)">/;
          var result = re.exec(desc);
          if(!result)
            return;

          // Remove sd's existing children
          sd.textContent = "";

          // Everything up to the link should be text content
          sd.appendChild(document.createTextNode(desc.slice(0, result.index)));

          // Now create the link itself
          var anchorEl = document.createElement("a");
          anchorEl.setAttribute("id", "cert_domain_link");
          anchorEl.setAttribute("title", result[1]);
          anchorEl.appendChild(document.createTextNode(result[1]));
          sd.appendChild(anchorEl);

          // Finally, append text for anything after the closing </a>
          sd.appendChild(document.createTextNode(desc.slice(desc.indexOf("</a>") + "</a>".length)));
        }

        var link = document.getElementById('cert_domain_link');
        if (!link)
          return;

        var okHost = link.getAttribute("title");
        var thisHost = document.location.hostname;
        var proto = document.location.protocol;

        // If okHost is a wildcard domain ("*.example.com") let's
        // use "www" instead.  "*.example.com" isn't going to
        // get anyone anywhere useful. bug 432491
        okHost = okHost.replace(/^\*\./, "www.");

        /* case #1:
         * example.com uses an invalid security certificate.
         *
         * The certificate is only valid for www.example.com
         *
         * Make sure to include the "." ahead of thisHost so that
         * a MitM attack on paypal.com doesn't hyperlink to "notpaypal.com"
         *
         * We'd normally just use a RegExp here except that we lack a
         * library function to escape them properly (bug 248062), and
         * domain names are famous for having '.' characters in them,
         * which would allow spurious and possibly hostile matches.
         */
        if (endsWith(okHost, "." + thisHost))
          link.href = proto + okHost;

        /* case #2:
         * browser.garage.maemo.org uses an invalid security certificate.
         *
         * The certificate is only valid for garage.maemo.org
         */
        if (endsWith(thisHost, "." + okHost))
          link.href = proto + okHost;
      }

      function endsWith(haystack, needle) {
        return haystack.slice(-needle.length) == needle;
      }

    ]]></script></head></html>
[info] [phantom] Step anonymous 2/2: done in 3824ms.
2017-08-05T02:38:37.581Z [DEBUG] network: request ignored. main page uri not started yet - jar:file:///usr/lib64/firefox/omni.ja!/chrome/toolkit/skin/classic/global/netError.css flags:TRANSFERRING,IS_REQ,
2017-08-05T02:38:37.582Z [DEBUG] network: progress total:0/0; uri: 2593/0 for jar:file:///usr/lib64/firefox/omni.ja!/chrome/toolkit/skin/classic/global/netError.css
2017-08-05T02:38:37.582Z [DEBUG] network: request ignored. main page uri not started yet - jar:file:///usr/lib64/firefox/omni.ja!/chrome/toolkit/skin/classic/global/netError.css flags:STOP,IS_REQ,
[info] [phantom] Done 2 steps in 3826ms
2017-08-05T02:38:37.604Z [DEBUG] webpage: close

```

### Expected results:

What is the correct way to use HTTPS proxy with slimerjs , since i assume i am using firefox so i have to use PAC file , and i did pass the correct valus , 

Please note the error : 
`2017-08-05T02:38:37.542Z [DEBUG] network: resource #1 response in error: #101 - The connection to the proxy server was refused`

Which is some thing related to how slimerjs send the request to the proxy server. i am not sure

Please advice

**Comment by laurentj** [comment]
Does it work [with this nightly](https://download.slimerjs.org/nightlies/1.0.0-alpha.2-pre/) ?

**Comment by laurentj** [comment]
No response. I close the issue. Reopen it if the problem is not resolved with the incoming SlimerJS 1.0
